components/swim/src/member.rs

```diff
@@ -0,0 +1,259 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::fmt;
+use std::iter::IntoIterator;
+use std::net::SocketAddr;
+use std::ops::{Deref, DerefMut};
+
+use uuid::Uuid;
+use rand::{thread_rng, Rng};
+
+use message::swim::Member as ProtoMember;
+
+const PINGREQ_TARGETS: usize = 5;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Health {
+    Alive,
+    Suspect,
+    Confirmed,
+}
+
+impl fmt::Display for Health {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Health::Alive => write!(f, "Alive"),
+            &Health::Suspect => write!(f, "Suspect"),
+            &Health::Confirmed => write!(f, "Confirmed"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Member {
+    pub proto: ProtoMember,
+}
+
+impl Member {
+    pub fn new() -> Member {
+        let mut proto_member = ProtoMember::new();
+        proto_member.set_id(Uuid::new_v4().simple().to_string());
+        proto_member.set_incarnation(0);
+        Member{
+            proto: proto_member,
+        }
+    }
+
+    pub fn new_from_proto(proto: ProtoMember) -> Member {
+        Member {
+            proto: proto,
+        }
+    }
+
+    pub fn socket_address(&self) -> SocketAddr {
+        match self.get_address().parse() {
+            Ok(addr) => addr,
+            Err(e) => {
+                panic!("Cannot parse member {:?} address: {}", self, e);
+            }
+        }
+    }
+}
+
+impl Deref for Member {
+    type Target = ProtoMember;
+
+    fn deref(&self) -> &ProtoMember {
+        &self.proto
+    }
+}
+
+impl DerefMut for Member {
+    fn deref_mut(&mut self) -> &mut ProtoMember {
+        &mut self.proto
+    }
+}
+
+
+// This is a Uuid type turned to a string
+pub type UuidSimple = String;
+
+#[derive(Debug, Clone)]
+pub struct MemberList {
+    members: HashMap<UuidSimple, Member>,
+    health: HashMap<UuidSimple, Health>
+}
+
+impl MemberList {
+    pub fn new() -> MemberList {
+        MemberList{
+            members: HashMap::new(),
+            health: HashMap::new(),
+        }
+    }
+
+    pub fn insert(&mut self, member: Member, health: Health) -> Option<Member> {
+        self.health.insert(String::from(member.get_id()), health);
+        self.members.insert(String::from(member.get_id()), member)
+    }
+
+    pub fn health_of(&self, member: &Member) -> Option<&Health> {
+        self.health.get(member.get_id())
+    }
+
+    pub fn insert_health(&mut self, member: &Member, health: Health) -> Option<Health> {
+        self.health.insert(String::from(member.get_id()), health)
+    }
+
+    pub fn members(&self) -> Vec<&Member> {
+        self.members.values().collect()
+    }
+
+    pub fn check_list(&self) -> Vec<Member> {
```

Comment:
I don't think you need to allocate a new list of Members here, just a ref should be alright by looking at the callers of `check_list()`. You can get a mutable member list with `values_mut()` to pass to `thread_rng::shuffle()` if that's what the motivation was.